src/react-components/chat-message.js

```diff
@@ -60,38 +94,123 @@ export function spawnChatMessage(body) {
         [styles.presenceLogEmoji]: isEmoji
       })}
     >
-      {messageBodyDom(body)}
+      {messageBodyDom(body, from)}
     </div>
   );
 
-  ReactDOM.render(entryDom, el, () => {
-    // Scale by 12x
-    messageCanvas.width = el.offsetWidth * 12.1;
-    messageCanvas.height = el.offsetHeight * 12.1;
-
-    const xhtml = encodeURIComponent(`
-        <svg xmlns="http://www.w3.org/2000/svg" width="${messageCanvas.width}" height="${messageCanvas.height}">
-          <foreignObject width="8.333%" height="8.333%" style="transform: scale(12.0);">
-            ${serializeElement(el)}
-          </foreignObject>
-        </svg>
-  `);
-    const img = new Image();
-
-    img.onload = async () => {
-      context.drawImage(img, 0, 0);
-      const blob = await new Promise(resolve => messageCanvas.toBlob(resolve));
-      document.querySelector("a-scene").emit("add_media", new File([blob], "message.png", { type: "image/png" }));
-      el.parentNode.removeChild(el);
-    };
-
-    img.src = "data:image/svg+xml," + xhtml;
+  return new Promise(resolve => {
+    ReactDOM.render(entryDom, el, () => {
+      // Scale by 12x
+      let objectScale = "8.33%";
+      let scale = 12;
+
+      if (lowResolution) {
+        // In low res, scale by 4x
+        objectScale = "25%";
+        scale = 4;
+      }
+
+      messageCanvas.width = el.offsetWidth * (scale + 0.1);
+      messageCanvas.height = el.offsetHeight * (scale + 0.1);
+
+      const xhtml = encodeURIComponent(`
+          <svg xmlns="http://www.w3.org/2000/svg" width="${messageCanvas.width}" height="${messageCanvas.height}">
+            <foreignObject width="${objectScale}" height="${objectScale}" style="transform: scale(${scale});">
+              ${serializeElement(el)}
+            </foreignObject>
+          </svg>
+    `);
+      const img = new Image();
+
+      img.onload = async () => {
+        context.drawImage(img, 0, 0);
+        const blob = await new Promise(resolve => messageCanvas.toBlob(resolve));
+        el.parentNode.removeChild(el);
+        resolve(blob);
+      };
+
+      img.src = "data:image/svg+xml," + xhtml;
+    });
   });
 }
 
-export default function ChatMessage(props) {
-  const isOneLine = props.body.split("\n").length === 1;
+export async function createInWorldLogMessage({ name, type, body }) {
+  if (type !== "chat") return;
+
+  const lowResolution = AFRAME.utils.device.isMobile();
+  const blob = await renderChatMessage(body, name, false, lowResolution);
+  const entity = document.createElement("a-entity");
+  const meshEntity = document.createElement("a-entity");
```

Comment:
Could we be reusing any resources here? Not sure whether perf is sensitive enough to merit a pool of entities (or even a pool of animations, if they are reusable).

Reply:
I think that'd just reduce memory churn, the overall perf impact of this stuff is primarily in the rendering step. I'd be nervous to rig that up at this point unless it proves a real problem, since it'd be easy to get wrong. I know we've tried to avoid memory allocations in behaviors that run each tick, but these allocations only happen when a chat message comes in so I think it's going to be OK for now.